--- ntfs-3g/include/ntfs-3g/logging.h
+++ ntfs-3g/include/ntfs-3g/logging.h
@@ -90,6 +90,7 @@ int ntfs_log_redirect(const char *functi
 /* Macros to simplify logging.  One for each level defined above.
  * Note, ntfs_log_debug/trace have effect only if DEBUG is defined.
  */
+#ifdef NO_OPTSIZE
 #define ntfs_log_critical(FORMAT, ARGS...) ntfs_log_redirect(__FUNCTION__,__FILE__,__LINE__,NTFS_LOG_LEVEL_CRITICAL,NULL,FORMAT,##ARGS)
 #define ntfs_log_error(FORMAT, ARGS...) ntfs_log_redirect(__FUNCTION__,__FILE__,__LINE__,NTFS_LOG_LEVEL_ERROR,NULL,FORMAT,##ARGS)
 #define ntfs_log_info(FORMAT, ARGS...) ntfs_log_redirect(__FUNCTION__,__FILE__,__LINE__,NTFS_LOG_LEVEL_INFO,NULL,FORMAT,##ARGS)
@@ -98,6 +99,16 @@ int ntfs_log_redirect(const char *functi
 #define ntfs_log_quiet(FORMAT, ARGS...) ntfs_log_redirect(__FUNCTION__,__FILE__,__LINE__,NTFS_LOG_LEVEL_QUIET,NULL,FORMAT,##ARGS)
 #define ntfs_log_verbose(FORMAT, ARGS...) ntfs_log_redirect(__FUNCTION__,__FILE__,__LINE__,NTFS_LOG_LEVEL_VERBOSE,NULL,FORMAT,##ARGS)
 #define ntfs_log_warning(FORMAT, ARGS...) ntfs_log_redirect(__FUNCTION__,__FILE__,__LINE__,NTFS_LOG_LEVEL_WARNING,NULL,FORMAT,##ARGS)
+#else
+#define ntfs_log_critical(FORMAT, args...) do { } while(0)
+#define ntfs_log_error(FORMAT, args...) do { } while(0)
+#define ntfs_log_info(FORMAT, args...) do { } while(0)
+#define ntfs_log_perror(FORMAT, args...) do { } while(0)
+#define ntfs_log_progress(FORMAT, args...) do { } while(0)
+#define ntfs_log_quiet(FORMAT, args...) do { } while(0)
+#define ntfs_log_verbose(FORMAT, args...) do { } while(0)
+#define ntfs_log_warning(FORMAT, args...) do { } while(0)
+#endif /* NO_OPTSIZE */
 
 /* By default debug and trace messages are compiled into the program,
  * but not displayed.
