--- dnsmasq/src/dhcp.c
+++ dnsmasq/src/dhcp.c
@@ -148,7 +148,9 @@
   time_t recvtime = now;
 #ifdef HAVE_LINUX_NETWORK
   struct arpreq arp_req;
+#ifndef HAVE_TOMATO
   struct timeval tv;
+#endif
   struct in_addr dst_addr;
 #endif
   
@@ -176,8 +178,10 @@
     return;
   
 #if defined (HAVE_LINUX_NETWORK)
+#ifndef HAVE_TOMATO
   if (ioctl(fd, SIOCGSTAMP, &tv) == 0)
     recvtime = tv.tv_sec;
+#endif
 
   dst_addr.s_addr = 0;
   
