--- dnsmasq/src/ipset.c
+++ dnsmasq/src/ipset.c
@@ -214,7 +214,7 @@ int add_to_ipset(const char *setname, co
   if (ret != -1) 
     ret = old_kernel ? old_add_to_ipset(setname, ipaddr, remove) : new_add_to_ipset(setname, ipaddr, af, remove);
 
-  if (ret == -1)
+  if (ret == -1 && errno != EEXIST && errno != EAFNOSUPPORT)
      my_syslog(LOG_ERR, _("failed to update ipset %s: %s"), setname, strerror(errno));
 
   return ret;
