--- miniupnpd/configure
+++ miniupnpd/configure
@@ -220,11 +220,12 @@
 	OS_VERSION=$(cat ./os.astlinux)
 fi
 
-# Tomato USB special case
+# FreshTomato special case
 if [ -f ../shared/tomato_version ]; then
 	OS_NAME=Tomato
-	TOMATO_VER=`cat ../shared/tomato_version | cut -d' ' -f2,3`
-	OS_VERSION="Tomato $TOMATO_VER"
+	TOMATO_VER=`cat ../shared/tomato_version | cut -d' ' -f1,2`
+	OS_VERSION="FreshTomato $TOMATO_VER"
+	PKG_CONFIG=
 fi
 
 # OpenEmbedded special case
@@ -557,19 +558,18 @@
 		;;
 	Tomato)
 		OS_FAMILY=Linux
-		OS_NAME=UPnP
-		OS_URL=http://tomatousb.org/
+		OS_NAME=FreshTomato
+		OS_URL=https://freshtomato.org/
 		echo "" >> ${CONFIGFILE}
 		echo "#ifndef TOMATO" >> ${CONFIGFILE}
 		echo "#define TOMATO" >> ${CONFIGFILE}
 		echo "#endif" >> ${CONFIGFILE}
 		echo "" >> ${CONFIGFILE}
-		echo "#ifdef LINUX26" >> ${CONFIGFILE}
 		echo "#define USE_IFACEWATCHER 1" >> ${CONFIGFILE}
-		echo "#endif" >> ${CONFIGFILE}
 		echo "#ifdef TCONFIG_IPV6" >> ${CONFIGFILE}
 		echo "#define ENABLE_IPV6" >> ${CONFIGFILE}
 		echo "#endif" >> ${CONFIGFILE}
+		HAVE_IP_MREQN=1
 		FW=iptables
 		;;
 	Darwin)
@@ -662,8 +662,14 @@
 				fi
 				if [ "$IPTABLES_143" = "1" ] ; then
 					echo "LDLIBS += ${IPTABLESPATH}/libiptc/.libs/libip4tc.o" >> ${CONFIG_MK}
+					if [ "$OS_NAME" = "FreshTomato" ] && [ "$TCONFIG_IPV6" = "y" ]; then
+						echo "LDLIBS += ${IPTABLESPATH}/libiptc/.libs/libip6tc.o" >> ${CONFIG_MK}
+					fi
 				else
 					echo "LDLIBS += ${IPTABLESPATH}/libiptc/libiptc.a" >> ${CONFIG_MK}
+					if [ "$OS_NAME" = "FreshTomato" ] && [ "$TCONFIG_IPV6" = "y" ]; then
+						echo "LDLIBS += ${IPTABLESPATH}/libiptc/libip6tc.a" >> ${CONFIG_MK}
+					fi
 				fi
 			else
 				# OpenWrt
--- miniupnpd/Makefile.linux
+++ miniupnpd/Makefile.linux
@@ -37,7 +37,7 @@
 CFLAGS ?= -Os
 CFLAGS += -fno-strict-aliasing
 CFLAGS += -fno-common
-CFLAGS += -fstack-protector -fPIE
+#CFLAGS += -fstack-protector -fPIE
 CFLAGS += -Wall
 CFLAGS += -Wextra -Wstrict-prototypes -Wdeclaration-after-statement
 #CFLAGS += -Wno-missing-field-initializers
@@ -48,7 +48,7 @@
 INSTALL = install
 STRIP ?= strip
 CP = cp
-DOXYGEN ?= doxygen
+#DOXYGEN ?= doxygen
 
 DEPDIR := .depend
 DEPFLAGS = -MMD -MP -MF $(@:%.o=$(DEPDIR)/%.d) -MT $@
@@ -65,7 +65,7 @@
 SYSTEMDUNITDIR = $(LIBDIR)/systemd/system
 
 include config.mk
-include $(SRCDIR)/gitrev.mk
+#include $(SRCDIR)/gitrev.mk
 include $(SRCDIR)/objects.mk
 
 # sources in netfilter/ directory
@@ -80,7 +80,7 @@
 
 #LDLIBS += -lnfnetlink
 
-EXECUTABLES = miniupnpd miniupnpdctl
+EXECUTABLES = miniupnpd
 
 ifeq ($(TESTS), 1)
 EXECUTABLES += testupnpdescgen testgetifstats \
--- miniupnpd/netfilter/iptcrdr.c
+++ miniupnpd/netfilter/iptcrdr.c
@@ -14,8 +14,12 @@
 #include <netinet/in.h>
 #include <arpa/inet.h>
 #include <dlfcn.h>
+#ifdef BCMARM
 #include <xtables.h>
+#else
+#include <iptables.h>
+#endif
 #include <linux/netfilter/xt_DSCP.h>
 #include <libiptc/libiptc.h>
 
 #include <linux/version.h>
@@ -33,7 +37,7 @@
 #define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]) + __must_be_array(arr))
 #define LIST_POISON2  ((void *) 0x00200200 )
 
-#if 0
+#ifndef BCMARM
 #include <linux/netfilter/nf_nat.h>
 #else
 #include "tiny_nf_nat.h"
@@ -46,7 +46,7 @@
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,22)
 #include <linux/netfilter_ipv4/ip_nat.h>
 #else
-#if 0
+#ifndef BCMARM
 #include <linux/netfilter/nf_nat.h>
 #else
 #include "tiny_nf_nat.h"
--- miniupnpd/netfilter/iptpinhole.c
+++ miniupnpd/netfilter/iptpinhole.c
@@ -20,7 +20,11 @@
 
 #ifdef ENABLE_UPNPPINHOLE
 
+#ifdef BCMARM
 #include <xtables.h>
+#else
+#include <iptables.h>
+#endif
 #include <libiptc/libip6tc.h>
 #include "tiny_nf_nat.h"
 
--- miniupnpd/netfilter/iptpinhole.c
+++ miniupnpd/netfilter/iptpinhole.c
@@ -174,7 +174,11 @@
                          struct ip6t_entry * e)
 {
 	IP6TC_HANDLE h;
+#ifdef BCMARM
 	xt_chainlabel chainlabel;
+#else
+	ipt_chainlabel chainlabel;
+#endif
 
 	h = ip6tc_init(table);
 	if(!h) {
@@ -331,7 +335,11 @@
 	const struct ip6t_entry_match *match = NULL;
 	/*const struct ip6t_entry_target *target = NULL;*/
 	unsigned int index;
+#ifdef BCMARM
 	xt_chainlabel chainlabel;
+#else
+	ipt_chainlabel chainlabel;
+#endif
 
 	p = get_pinhole(uid);
 	if(!p)
--- miniupnpd/testifacewatcher.c
+++ miniupnpd/testifacewatcher.c
@@ -2,6 +2,7 @@
 
 #include <syslog.h>
 #include <signal.h>
+#include <stdio.h>
 
 #include "ifacewatcher.h"
 #include "miniupnpdtypes.h"
