--- miniupnpd/genconfig.sh
+++ miniupnpd/genconfig.sh
@@ -94,7 +94,7 @@ fi
 # Tomato USB special case
 if [ -f ../shared/tomato_version ]; then
 	OS_NAME=Tomato
-	TOMATO_VER=`cat ../shared/tomato_version | cut -d' ' -f2,3`
+	TOMATO_VER=`cat ../shared/tomato_version | cut -d' ' -f1,2`
 	OS_VERSION="Tomato $TOMATO_VER"
 fi
 
--- miniupnpd/Makefile.linux
+++ miniupnpd/Makefile.linux
@@ -27,7 +27,7 @@ CFLAGS ?= -Os
 CFLAGS += -fno-common
 CFLAGS += -fstack-protector -fPIE
 CFLAGS += -D_FORTIFY_SOURCE=2
-CPPFLAGS += -D_GNU_SOURCE
+CFLAGS += -D_GNU_SOURCE
 CFLAGS += -Wall
 CFLAGS += -Wextra -Wstrict-prototypes -Wdeclaration-after-statement
 #CFLAGS += -Wno-missing-field-initializers
@@ -162,7 +162,7 @@ LDLIBS += $(shell $(PKG_CONFIG) --static
 
 # OpenWrt packager disables https server for IGD v2 and hardcodes libuuid support
 ifeq ($(TARGET_OPENWRT),)
-LDLIBS += $(shell $(PKG_CONFIG) --static --libs-only-l libssl)
+#LDLIBS += $(shell $(PKG_CONFIG) --static --libs-only-l libssl)
 
 TEST := $(shell $(PKG_CONFIG) --exists uuid && echo 1)
 ifeq ($(TEST),1)
--- miniupnpd/netfilter/iptcrdr.c
+++ miniupnpd/netfilter/iptcrdr.c
@@ -14,8 +14,8 @@
 #include <netinet/in.h>
 #include <arpa/inet.h>
 #include <dlfcn.h>
-#include <xtables.h>
+#include <iptables.h>
 #include <linux/netfilter/xt_DSCP.h>
 #include <libiptc/libiptc.h>
 
 #include <linux/version.h>
@@ -31,8 +31,8 @@
 #define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]) + __must_be_array(arr))
 #define LIST_POISON2  ((void *) 0x00200200 )
 
-#if 0
-#include <linux/netfilter/nf_nat.h>
+#if 1
+#include "../../iptables-1.4.x/include/net/netfilter/nf_nat.h"
 #else
 #include "tiny_nf_nat.h"
 #endif
@@ -44,7 +44,7 @@
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,22)
 #include <linux/netfilter_ipv4/ip_nat.h>
 #else
-#if 0
+#if 1
 #include <linux/netfilter/nf_nat.h>
 #else
 #include "tiny_nf_nat.h"
