--- miniupnpd/Makefile.linux
+++ miniupnpd/Makefile.linux
@@ -120,7 +120,13 @@ TEST := $(shell [ $(IPTVER1) -gt 1 ] ||
 LDLIBS = $(IPTABLESPATH)/libiptc/.libs/libip4tc.o
 #LDLIBS = $(IPTABLESPATH)/libiptc/.libs/libiptc.a
 else # ifeq ($(TEST), 1)
+# Tomato - check IPv6 support
+ifeq ($(TCONFIG_IPV6),y)
+LDLIBS = $(IPTABLESPATH)/libiptc/libiptc.a
+LDLIBS += $(IPTABLESPATH)/libiptc/libip6tc.a
+else # ($(TCONFIG_IPV6),y)
 LDLIBS = $(IPTABLESPATH)/libiptc/libiptc.a
+endif # ($(TCONFIG_IPV6),y)
 endif # ifeq ($(TEST), 1)
 else # ($(TARGET_OPENWRT),)
 # openWRT :
--- miniupnpd/netfilter/iptpinhole.c
+++ miniupnpd/netfilter/iptpinhole.c
@@ -20,7 +20,7 @@
 
 #ifdef ENABLE_UPNPPINHOLE
 
-#include <xtables.h>
+#include <iptables.h>
 #include <libiptc/libip6tc.h>
 #include "tiny_nf_nat.h"
 
