--- ffmpeg/libavutil/log.c
+++ ffmpeg/libavutil/log.c
@@ -88,6 +88,7 @@ void av_log_default_callback(void* ptr,
 
 static void (*av_log_callback)(void*, int, const char*, va_list) = av_log_default_callback;
 
+#ifdef NO_OPTSIZE
 void av_log(void* avcl, int level, const char *fmt, ...)
 {
     va_list vl;
@@ -100,6 +101,7 @@ void av_vlog(void* avcl, int level, cons
 {
     av_log_callback(avcl, level, fmt, vl);
 }
+#endif /* NO_OPTSIZE */
 
 int av_log_get_level(void)
 {
--- ffmpeg/libavutil/log.h
+++ ffmpeg/libavutil/log.h
@@ -108,6 +108,7 @@ typedef struct {
  * subsequent arguments are converted to output.
  * @see av_vlog
  */
+#ifdef NO_OPTSIZE
 #ifdef __GNUC__
 void av_log(void*, int level, const char *fmt, ...) __attribute__ ((__format__ (__printf__, 3, 4)));
 #else
@@ -115,6 +116,11 @@ void av_log(void*, int level, const char
 #endif
 
 void av_vlog(void*, int level, const char *fmt, va_list);
+#else
+#define av_log(acvl, level, fmt, ...) do { } while(0)
+#define av_vlog(avcl, level, fmt, va_list) do { } while(0)
+#endif /* NO_OPTSIZE */
+
 int av_log_get_level(void);
 void av_log_set_level(int);
 void av_log_set_callback(void (*)(void*, int, const char*, va_list));
