diff -Naurp openvpn/src/openvpn/route.c openvpn/src/openvpn/route.c
--- openvpn/src/openvpn/route.c	2017-03-22 16:34:21.000000000 +0100
+++ openvpn/src/openvpn/route.c	2018-04-17 17:15:46.759443944 +0200
@@ -45,6 +45,10 @@
 #include "memdbg.h"
 
 #if defined(TARGET_LINUX) || defined(TARGET_ANDROID)
+// be compatible with Tomato K26 - shibby
+typedef __signed__ long __s64;
+typedef unsigned long __u64;
+// end - shibby
 #include <linux/rtnetlink.h>            /* RTM_GETROUTE etc. */
 #endif
 
diff -Naurp openvpn/src/openvpn/tun.c openvpn/src/openvpn/tun.c
--- openvpn/src/openvpn/tun.c	2017-03-22 16:34:21.000000000 +0100
+++ openvpn/src/openvpn/tun.c	2018-04-17 17:15:46.767444024 +0200
@@ -2054,6 +2054,20 @@ open_tun(const char *dev, const char *de
 
 #ifdef ENABLE_FEATURE_TUN_PERSIST
 
+/*
+ * This can be removed in future
+ * when all systems will use newer
+ * linux-headers
+ *
+ * be compatible with Tomato K26 - shibby
+ */
+#ifndef TUNSETOWNER
+#define TUNSETOWNER	_IOW('T', 204, int)
+#endif
+#ifndef TUNSETGROUP
+#define TUNSETGROUP	_IOW('T', 206, int)
+#endif
+
 void
 tuncfg(const char *dev, const char *dev_type, const char *dev_node, int persist_mode, const char *username, const char *groupname, const struct tuntap_options *options)
 {
