--- dropbear/configure.ac
+++ dropbear/configure.ac
@@ -116,22 +116,6 @@ if test "$hardenbuild" -eq 1; then
 			[AC_MSG_NOTICE([Not setting $TESTFLAGS]); LDFLAGS="$OLDLDFLAGS" ]
 			)
 	fi # non-static
-	# stack protector. -strong is good but only in gcc 4.9 or later
-	OLDCFLAGS="$CFLAGS"
-	TESTFLAGS="-fstack-protector-strong"
-	CFLAGS="$TESTFLAGS $CFLAGS"
-	AC_LINK_IFELSE([AC_LANG_PROGRAM([])],
-	    [AC_MSG_NOTICE([Setting $TESTFLAGS])],
-	    [
-			CFLAGS="$OLDCFLAGS"
-			TESTFLAGS="-fstack-protector --param=ssp-buffer-size=4"
-			CFLAGS="$TESTFLAGS $CFLAGS"
-			AC_LINK_IFELSE([AC_LANG_PROGRAM([])],
-			    [AC_MSG_NOTICE([Setting $TESTFLAGS])],
-			    [AC_MSG_NOTICE([Not setting $TESTFLAGS]); CFLAGS="$OLDCFLAGS" ]
-			    )
-	    ]
-	    )
 	# FORTIFY_SOURCE
 	DB_TRYADDCFLAGS([-D_FORTIFY_SOURCE=2])
 
--- dropbear/libtommath/Makefile.in
+++ dropbear/libtommath/Makefile.in
@@ -7,7 +7,7 @@ srcdir=@srcdir@
 
 # So that libtommath can include Dropbear headers for options and m_burn()
 CFLAGS += -I$(srcdir) -I../libtomcrypt/src/headers/ -I$(srcdir)/../libtomcrypt/src/headers/ -I../ -I$(srcdir)/../
-CFLAGS += -Wno-deprecated
+#CFLAGS += -Wno-deprecated
 CFLAGS += $(CPPFLAGS)
 
 V = 1
