--- nettle/configure
+++ nettle/configure
@@ -6060,10 +6060,10 @@ fi
 $as_echo "$nettle_cv_c_builtin_bswap64" >&6; }
 
 
-if test "x$nettle_cv_c_builtin_bswap64" = "xyes" ; then
-  $as_echo "#define HAVE_BUILTIN_BSWAP64 1" >>confdefs.h
-
-fi
+#if test "x$nettle_cv_c_builtin_bswap64" = "xyes" ; then
+#  $as_echo "#define HAVE_BUILTIN_BSWAP64 1" >>confdefs.h
+#
+#fi
 
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for __attribute__" >&5
 $as_echo_n "checking for __attribute__... " >&6; }
