--- pppd/pppd/Makefile.linux
+++ pppd/pppd/Makefile.linux
@@ -32,7 +32,7 @@ endif
 
 # CC = gcc
 #
-COPTS = -O2 -pipe -Wall -g
+COPTS = -Os -pipe -Wall -g -ffunction-sections -fdata-sections
 LIBS = -lrt
 
 # Uncomment the next line to include support for Microsoft's
@@ -183,7 +183,7 @@ endif
 
 ifdef PLUGIN
 CFLAGS	+= -DPLUGIN
-LDFLAGS_PLUGIN	+= -Wl,-E
+LDFLAGS_PLUGIN	+= -Wl,-E -ffunction-sections -fdata-sections -Wl,--gc-sections
 LIBS	+= -ldl
 endif
 
--- pppd/pppd/plugins/pppol2tp/Makefile.linux
+++ pppd/pppd/plugins/pppol2tp/Makefile.linux
@@ -1,5 +1,5 @@
 #CC	= gcc
-COPTS	= -O2 -g
+COPTS	= -Os -g -ffunction-sections -fdata-sections
 CFLAGS	= $(COPTS) -I. -I../.. -I../../../include -fPIC
 LDFLAGS	= -shared
 INSTALL	= install
--- pppd/chat/Makefile.linux
+++ pppd/chat/Makefile.linux
@@ -10,7 +10,7 @@ CDEF3=	-UNO_SLEEP			# Use the usleep fun
 CDEF4=	-DFNDELAY=O_NDELAY		# Old name value
 CDEFS=	$(CDEF1) $(CDEF2) $(CDEF3) $(CDEF4)
 
-COPTS=	-O2 -g -pipe
+COPTS=	-Os -g -pipe -ffunction-sections -fdata-sections
 CFLAGS=	$(COPTS) $(CDEFS)
 
 INSTALL= install
