--- libubox/CMakeLists.txt
+++ libubox/CMakeLists.txt
@@ -15,8 +15,8 @@
 OPTION(BUILD_EXAMPLES "build examples" ON)
 
 INCLUDE(FindPkgConfig)
-PKG_SEARCH_MODULE(JSONC json-c REQUIRED)
-INCLUDE_DIRECTORIES(${JSONC_INCLUDE_DIRS})
+#PKG_SEARCH_MODULE(JSONC json-c REQUIRED)
+#INCLUDE_DIRECTORIES(${JSONC_INCLUDE_DIRS})
 
 SET(SOURCES avl.c avl-cmp.c blob.c blobmsg.c uloop.c usock.c ustream.c ustream-fd.c vlist.c utils.c safe_list.c runqueue.c md5.c kvlist.c ulog.c base64.c udebug.c udebug-remote.c)
 
--- libubox/blobmsg_json.c
+++ libubox/blobmsg_json.c
@@ -17,7 +17,7 @@
 #include "blobmsg.h"
 #include "blobmsg_json.h"
 
-#include <json.h>
+#include <json-c/json.h>
 
 bool blobmsg_add_object(struct blob_buf *b, json_object *obj)
 {
--- libubox/jshn.c
+++ libubox/jshn.c
@@ -13,7 +13,7 @@
  * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
  * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
  */
-#include <json.h>
+#include <json-c/json.h>
 #include <string.h>
 #include <stdlib.h>
 #include <stdio.h>
